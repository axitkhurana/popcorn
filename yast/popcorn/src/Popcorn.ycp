/* ------------------------------------------------------------------------------
 * Copyright (c) 2006 Novell, Inc. All Rights Reserved.
 *
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of version 2 of the GNU General Public License as published by the
 * Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, contact Novell, Inc.
 *
 * To contact Novell about this file by physical or electronic mail, you may find
 * current contact information at www.novell.com.
 * ------------------------------------------------------------------------------
 */

/**
 * File:	modules/Popcorn.ycp
 * Package:	Configuration of popcorn
 * Summary:	Popcorn settings, input and output functions
 * Authors:	Akshit Khurana <axitkhurana@gmail.com>
 *
 * $Id: Popcorn.ycp 41350 2007-10-10 16:59:00Z dfiser $
 *
 * Representation of the configuration of popcorn.
 * Input and output routines.
 */

{

module "Popcorn";
textdomain "popcorn";

import "Progress";
import "Report";
import "Summary";
import "Message";
import "SCR";

/**
 * Prototypes
 */
global boolean Modified();

/**
 * Data was modified?
 */
boolean modified = false;

/**
 */
boolean proposal_valid = false;

/**
 * Write only, used during autoinstallation.
 * Don't run services and SuSEconfig, it's all done at one place.
 */
boolean write_only = false;

/**
 * Abort function
 * return boolean return true if abort
 */
boolean() AbortFunction = Modified;

/**
 * Abort function
 * @return boolean return true if abort
 */
global define boolean Abort() ``{
    if(AbortFunction != nil)
    {
	return AbortFunction () == true;
    }
    return false;
}

/**
 * Data was modified?
 * @return true if modified
 */
global boolean Modified() {
    y2debug("modified=%1",modified);
    return modified;
}

/**
 * Mark as modified, for Autoyast.
 */
global void SetModified(boolean value) {
      modified = true;
  }

global boolean ProposalValid() {
    return proposal_valid;
}

global void SetProposalValid(boolean value) {
    proposal_valid = value;
}

/**
 * @return true if module is marked as "write only" (don't start services etc...)
 */
global boolean WriteOnly() {
    return write_only;
}

/**
 * Set write_only flag (for autoinstalation).
 */
global void SetWriteOnly(boolean value) {
    write_only = value;
}


global void SetAbortFunction(boolean() function) {
    AbortFunction = function;
}

typedef map<string, any> ConfigEntry;

// Settings: Variables needed for configuration of popcorn
global list<ConfigEntry> popcorn_config =
    [
     $[ "name": "popcorn_enabled", "type": "bool", "def": true ],
    ];

global path popcorn_config_path = .popcorn.value.settings;

// Mapping from config entry names to widget captions
// (in a real life example this would go to the ConfigEntry map, too)

global map<string, string> widget_caption =
    $[
      "popcorn_enabled"     : "Enable Popcorn",
      ];


/**
 * Create a widget for a config entry
 *
 * Parameters:
 *     entry                map describing the config entry
 *     scr_base_path        SCR path to use (entry["name"] will be added)
 *
 * Return:
 *     term describing the widget
 **/
global term entryWidget( ConfigEntry entry, path scr_base_path )
{
    string name     = (string) entry[ "name" ]:nil;
    string type     = (string) entry[ "type" ]:"string";
    string caption  = widget_caption[ name ]:name;

    // Read current value from config file

    any value = SCR::Read( add( scr_base_path, name ) );
    y2debug("Read value %1", value);

    // Create corresponding widget

    term widget = nil;

    if ( type == "string" )
    {
        widget = `TextEntry( `id( name ), caption, value != nil ? value : "" );
    }
    else if ( type == "password" )
    {
        widget = `Password( `id( name ), caption, value != nil ? value : "" );
    }
    else if ( type == "bool" )
    {
        widget = `Left( `CheckBox( `id( name ), caption, value == "yes" ) );
    }
    else if ( type == "int" )
    {
        integer min = tointeger( entry[ "min" ]:0 );
        integer max = tointeger( entry[ "max" ]:65535 );
        widget = `IntField( `id( name ), caption, min, max,
                            value != nil ? tointeger( value ) : min );
    }
    else
    {
        y2error( "Unknown type in config entry: %1", entry );
    }

    return widget;
}


/**
 * Create widgets in a VBox for a list of config entries
 *
 * Parameters:
 *     scr_base_path        SCR path to use (entry["name"] will be added)
 *     entries              list of maps describing the config entries
 *
 * Return:
 *     widget term
 **/
global term configWidgets( list<ConfigEntry> entry_list, path scr_base_path )
{
    term vbox = `VBox();

    foreach( ConfigEntry entry, entry_list,
    {
        term widget = entryWidget( entry, scr_base_path );

        if ( widget != nil )
            vbox = add( vbox, widget );
    });

    return vbox;
}


/**
 * Write a list of configuration entries to file.
 * Get the current value for each entry from a widget in the current dialog.
 **/
global void writeConfig( list<ConfigEntry> entry_list, path scr_base_path )
{
    foreach( ConfigEntry entry, entry_list,
    {
        string name = (string) entry[ "name" ]:nil;

        if ( name == nil )
        {
            y2error( "Entry without name in entry list: %1", entry );
        }
        else
        {
            any value = UI::QueryWidget(`id( name ), `Value );

            if ( is( value, boolean ) )
                value = ( (boolean) value ) ? "yes" : "no";

            SCR::Write( add( scr_base_path, name ), value );
        }
    });
}


// XXX Remove unused functions

/**
 * Read all popcorn settings
 * @return true on success
 */
global boolean Read() {

    /* Popcorn read dialog caption */
    string caption = _("Initializing popcorn Configuration");

    // TODO FIXME Set the right number of stages
    integer steps = 4;

    integer sl = 500;
    sleep(sl);

    // TODO FIXME Names of real stages
    // We do not set help text here, because it was set outside
    Progress::New( caption, " ", steps, [
	    /* Progress stage 1/3 */
	    _("Read the database"),
	    /* Progress stage 2/3 */
	    _("Read the previous settings"),
	    /* Progress stage 3/3 */
	    _("Detect the devices")
	], [
	    /* Progress step 1/3 */
	    _("Reading the database..."),
	    /* Progress step 2/3 */
	    _("Reading the previous settings..."),
	    /* Progress step 3/3 */
	    _("Detecting the devices..."),
	    /* Progress finished */
	    _("Finished")
	],
	""
    );

    // read database
    if(Abort()) return false;
    Progress::NextStage();
    /* Error message */
    if(false) Report::Error(_("Cannot read database1."));
    sleep(sl);

    // read another database
    if(Abort()) return false;
    Progress::NextStep();
    /* Error message */
    if(false) Report::Error(_("Cannot read database2."));
    sleep(sl);

    // read current settings
    if(Abort()) return false;
    Progress::NextStage();
    /* Error message */
    if(false) Report::Error(Message::CannotReadCurrentSettings());
    sleep(sl);

    // detect devices
    if(Abort()) return false;
    Progress::NextStage();
    /* Error message */
    if(false) Report::Warning(_("Cannot detect devices."));
    sleep(sl);

    if(Abort()) return false;
    /* Progress finished */
    Progress::NextStage();
    sleep(sl);

    if(Abort()) return false;
    modified = false;
    return true;
}

/**
 * Write all popcorn settings
 * @return true on success
 */
global boolean Write() {

    /* Popcorn read dialog caption */
    string caption = _("Saving popcorn Configuration");

    // TODO FIXME And set the right number of stages
    integer steps = 2;

    integer sl = 500;
    sleep(sl);

    // TODO FIXME Names of real stages
    // We do not set help text here, because it was set outside
    Progress::New(caption, " ", steps, [
	    /* Progress stage 1/2 */
	    _("Write the settings"),
	    /* Progress stage 2/2 */
	    _("Run SuSEconfig")
	], [
	    /* Progress step 1/2 */
	    _("Writing the settings..."),
	    /* Progress step 2/2 */
	    _("Running SuSEconfig..."),
	    /* Progress finished */
	    _("Finished")
	],
	""
    );

    // write settings
    if(Abort()) return false;
    Progress::NextStage();
    /* Error message */
    if(false) Report::Error (_("Cannot write settings."));
    sleep(sl);

    // run SuSEconfig
    if(Abort()) return false;
    Progress::NextStage ();
    /* Error message */
    if(false) Report::Error (Message::SuSEConfigFailed());
    sleep(sl);

    if(Abort()) return false;
    /* Progress finished */
    Progress::NextStage();
    sleep(sl);

    if(Abort()) return false;
    return true;
}

/**
 * Get all popcorn settings from the first parameter
 * (For use by autoinstallation.)
 * @param settings The YCP structure to be imported.
 * @return boolean True on success
 */
global boolean Import (map settings) {
    // TODO FIXME: your code here (fill the above mentioned variables)...
    return true;
}

/**
 * Dump the popcorn settings to a single map
 * (For use by autoinstallation.)
 * @return map Dumped settings (later acceptable by Import ())
 */
global map Export () {
    // TODO FIXME: your code here (return the above mentioned variables)...
    return $[];
}

/**
 * Create a textual summary and a list of unconfigured cards
 * @return summary of the current configuration
 */
global list Summary() {
    // TODO FIXME: your code here...
    /* Configuration summary text for autoyast */
    return [ _("Configuration summary..."), [] ];
}

/**
 * Create an overview table with all configured cards
 * @return table items
 */
global list Overview() {
    // TODO FIXME: your code here...
    return [];
}

/**
 * Return packages needed to be installed and removed during
 * Autoinstallation to insure module has all needed software
 * installed.
 * @return map with 2 lists.
 */
global map AutoPackages() {
    // TODO FIXME: your code here...
    return $[ "install":[], "remove":[] ];
}

/* EOF */
}
