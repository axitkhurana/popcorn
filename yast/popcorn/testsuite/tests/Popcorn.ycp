{

// testedfiles: Popcorn.ycp

include "testsuite.ycp";
TESTSUITE_INIT([], nil);

import "Popcorn";

DUMP("Popcorn::Modified");
TEST(``(Popcorn::Modified()), [], nil);

}
