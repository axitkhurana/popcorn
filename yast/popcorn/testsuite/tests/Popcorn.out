Dump	Popcorn::Modified
Return	false
